### PR TITLE
2.x user factories

### PR DESCRIPTION
**Pull request by acobster**
**Ticket**: #2073 #2088 #2094 

## Issue

Updating the Users API per discussion in #2073.

## Solution

Extensive discussion of the solution is in #2073 and #2093. In short:

* `Timber::get_user` and `Timber::get_users()` no longer take a `$UserClass` arg; users must now rely on the `timber/user/classmap` filter to achieve polymorphic instantiation.
* The public Timber functions use Factories under the hood, which in turn apply `timber/user/classmap`.

## Impact

Discussed in docs.

## Usage Changes

Also discussed in the docs.

## Considerations

I'm not sure what the new usage of the Twig `{{ User(...) }}` fn should be. Currently it supports arrays of IDs, which seems a bit astonishing. I'd be in favor or dropping support for this, or at least deprecating that use-case and providing a new `{{ Users(...) }}` (or `{{ UserCollection(...) }}`?) function.

Speaking of which, do we need a `UserCollection` class? I kind of assumed in the back of my mind that we'd want Collections for Users and Terms (we already have `CommentThread`), but thinking more critically about it, the Loop is the main reason we have `PostCollection`, and there's not really an analog to that with Users or Terms. I could see there being a use-case for pagination or some other advanced query-related stuff, but I'm not sure. Thoughts?

**Edit:** oh, and I'm hoping one of you can look at #2157 before we merge this in. :)

## Testing

Tests have been refactored and removed where appropriate. The `UserFactory` class has some prety thorough tests as well. It would be great add to tests for any edge cases y'all can think of.

### CODE REVIEW

**Comment by gchtr** [comment]
> I'm not sure what the new usage of the Twig `{{ User(...) }}` fn should be. Currently it supports arrays of IDs, which seems a bit astonishing. I'd be in favor or dropping support for this, or at least deprecating that use-case and providing a new `{{ Users(...) }}` (or `{{ UserCollection(...) }}`?) function.

I think this was applied to match the behavior of other other types (post, terms, ...), without thinking about a use case.

But we will deprecate the Twig functions anyway and use the same function names that we use in PHP, see https://github.com/timber/timber/issues/2144. That would mean we should have a `{{ get_users(...) }}` function as well. The same goes for the other objects.

> Speaking of which, do we need a `UserCollection` class? I kind of assumed in the back of my mind that we'd want Collections for Users and Terms (we already have `CommentThread`), but thinking more critically about it, the Loop is the main reason we have `PostCollection`, and there's not really an analog to that with Users or Terms. I could see there being a use-case for pagination or some other advanced query-related stuff, but I'm not sure. Thoughts?

We discussed this quite some time ago in https://github.com/timber/timber/issues/1689. My take is that we don’t need UserCollections, because the loop isn’t involved and usually we don’t need pagination. The same goes for Terms. In case we ever need that in the future, we can add it later.

I will have a thorough look at this later this week.

**Comment by acobster** [comment]
@gchtr ok, I remember that discussion now. I guess we can always add a `::get_users()` option to return a `UserCollection` instead of an array if we decide to implement it. It also occurs to me that we should probably have a `users.md` among the [Guides](https://github.com/timber/timber/tree/2.x-docs-api/docs/guides). Do you have the bandwidth to write one up?

**Comment by gchtr** [comment]
👍

> It also occurs to me that we should probably have a `users.md` among the [Guides](https://github.com/timber/timber/tree/2.x-docs-api/docs/guides) [](/timber/timber/tree/HEAD@{2020-01-09T20:20:57Z}/docs/guides). Do you have the bandwidth to write one up?

Yes, I will write one up 👍

File: lib/Timber.php [hunk 1]
```diff
@@ -215,6 +216,65 @@ public static function get_term( $term, $taxonomy = 'post_tag', $TermClass = 'Ti
...
+	================================ */
+
+	/**
+	 * Get one or more users as a collection.
+	 * @api
+	 * @param array   $query The params for the WP_User_Query to performed
+	 * @param array   $options optional; none are currently supported
+	 * @return mixed
+	 */
+	public static function get_users( array $query = [], array $options = [] ) : Iterable {
```
Comment (gchtr) [inline lib/Timber.php]: Should we really add an `$options` parameter if there are no options supported yet?

File: lib/Factory/UserFactory.php [hunk 2]
```diff
@@ -35,10 +35,12 @@ public function from($params) {
...
 
 	protected function from_id(int $id) {
-		return $this->build(get_user_by('id', $id));
+		$wp_user = get_user_by('id', $id);
+
+		return $wp_user ? $this->build($wp_user) : null;
 	}
 
-	protected function from_post_obj(object $obj) : CoreInterface {
+	protected function from_user_object(object $obj) : CoreInterface {
```
Comment (gchtr) [inline lib/Factory/UserFactory.php]: The `object` type hint is only supported in PHP 7.2+. So I guess we can’t use it yet.

File: lib/Factory/UserFactory.php [hunk 3]
```diff
@@ -35,10 +35,12 @@ public function from($params) {
 	}
```
Comment (gchtr) [inline lib/Factory/UserFactory.php]: I’m getting a "Missing return statement" warning here.

Comment (acobster) [reply lib/Factory/UserFactory.php]: I added an explicit `return null` at the end of this method, but that's only a guess as to what this means. I'm not seeing that warning on my end. Are you getting it running `phpunit`?

File: lib/Factory/UserFactory.php (hunk 2 above)
Comment (acobster) [reply lib/Factory/UserFactory.php]: Removed.

File: lib/Timber.php (hunk 1 above)
Comment (acobster) [reply lib/Timber.php]: Not sure about this one. This was back when I thought we'd return a `UserCollection`, so adding such an option might be imminent. So not as necessary. However, it does still come down to what constitutes a "breaking change" and the kinds of use-cases we explicitly support.

If don't ship with `$options` but then add it later, then someone could literally do `MyClass extends Timber` and override this method with single arity on 2.0 and then encounter a breakage just by updating to a minor version when we add the param. So that's technically a breaking change, even if we can't imagine why someone would do that instead of just using the filters we provide. But as long as there's a possibility, I think it might be good to signal that intention and leave the door open.

Curious to hear from @jarednova here.

Comment (jarednova) [reply lib/Timber.php]: So it's basically like no options yet, but (perhaps) options to come? A good case could be made either way, but it seems like it could be helpful in remembering for when options get added this is where they go in order to match patterns seen elsewhere in the code base. Do I have that right?

Comment (acobster) [reply lib/Timber.php]: Yes.

File: lib/Factory/UserFactory.php (hunk 3 above)
Comment (gchtr) [reply lib/Factory/UserFactory.php]: I’m seeing this in PhpStorm as an [Inconsistent return points](https://www.jetbrains.com/help/phpstorm/php-inconsistent-return-points.html) warning.

I’m not sure what the best solution is here. A `@return mixed|void` would also make the warning go away, but should we ever return `void` if we have other returns in a method?

If other methods like `UserFactory::from_id()` return `null` when no `Timber\User` could be returned, wouldn’t it make sense to be explicit and return `null` here, too?

File: lib/Timber.php (hunk 1 above)
Comment (gchtr) [reply lib/Timber.php]: I didn’t think about the breaking change if someone extends `Timber\Timber`. Under these circumstances, it makes total sense. And even if we don’t have a use case now, there can very well be a use case in the future, as we have learned time and time again while developing and maintaining Timber.

Let’s leave it, then! 👍

File: lib/Factory/UserFactory.php (hunk 3 above)
Comment (acobster) [reply lib/Factory/UserFactory.php]: `::from_id` already returns explicitly...right? It just returns the result of the ternary expression.

I added some doc comments that I feel are appropriately specific. I don't care too much about appeasing PHPStorm since we're already [adding PHPStan](https://github.com/timber/timber/pull/2149) in parallel. Can't please everybody. :)

**Comment by acobster** [comment]
@gchtr I just got all unit tests passing for this locally.

Per discussion on #2157 I'm dropping support for Co-Authors Plus for now, as I just don't have time to continue supporting it. Discussion can continue on that PR, where the `2.x-coauthors-fixes` branch contains all the code needed to re-add support for it.

How are you feeling about this PR?

File: lib/Timber.php [hunk 4]
```diff
@@ -215,6 +216,65 @@ public static function get_term( $term, $taxonomy = 'post_tag', $TermClass = 'Ti
...
+	 * - email
+	 * - login
+	 * @param int|string $value The value for $field
+	 * @return \Timber\User|null
+	 */
+	public static function get_user_by( string $field, $value ) {
+		$wp_user = get_user_by($field, $value);
+
+		if ($wp_user === false) {
+			return false;
```
Comment (jarednova) [inline lib/Timber.php]: Just wanted to note that I changed this from returning `null` to `false`. It'd seem to me that if no user is found `false` is a more logical representation

Comment (gchtr) [reply lib/Timber.php]: @jarednova To me, this is a question of: What do we return in any of the `Timber::get_*` functions when no object could be found based on the input. Is it `null` or `false`? And does it matter?

**Comment by jarednova** [comment]
Extensive work here @acobster! I found a few tiny things, but overall this is looking super. Take a look and let me know if you agree with my one substantive change (returning `false` instead of `null` when no user found via `Timber::get_user_by` and other similar methods). This is the pattern I saw in `CommentFactory` and what would make the most logical sense to me*

Merge if you agree! (or we can discuss what makes the most sense here and close that out)

`*` I realize there are a few other places in the 2.x codebase not related to factories where things are returning `null` instead of `false`. I'll take a separate PR/issue to investigate and resolve these

**Comment by jarednova** [comment]
... and also to close the loop on CoAuthorsPlus. I think #2157 is the perfect place to either resolve or abandon attempting to make it compatible. Let's not let that block us here

**Comment by acobster** [comment]
@jarednova yeah, I'm assuming the magic Timber getters will still return `false` when they don't find anything? In keeping with that I think it makes sense to return `false` here too.

Thanks for adding those tests, looking good :+1:

**Comment by jarednova** [comment]
Awesome, feels great to get this merged!

**Comment by jarednova** [comment]
Awesome, feels great to get this merged!
